### PR TITLE
Fix #20581: preserve staff speed over cheat window open/close cycles

### LINKED ISSUE #20581: Staff speed setting reverts to display "Normal" when menu is re-opened
### Operating System

MacOS Ventura 13.2.1

### OpenRCT2 build

OpenRCT2, v0.4.5 (76ca840 on HEAD)

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

_No response_

### Describe the issue

Staff speed setting reverts to displaying "Normal" from "Fast" or "Frozen" when the park cheats tab in the cheats menu is opened. This is only a display bug -- the actual speed of staff is unaffected until you interact with the option.

### Steps to reproduce

1. Open cheat menu
2. Navigate to "Park Cheats" tab
3. Modify the "Staff speed" option to "Fast" or "Frozen" (from its default state of "Normal")
4. Close the cheat menu
5. Open the cheat menu
6. Navigate to "Park Cheats" tab

The "Staff speed" option will have reverted to displaying its default state of "Normal", rather than its true state, which may still be "Fast" or "Frozen".

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by KawkMob**

### CODE REVIEW

File: src/openrct2-ui/windows/Cheats.cpp [hunk 1]
```diff
@@ -354,14 +354,15 @@ static StringId window_cheats_page_titles[] = {
     STR_CHEAT_TITLE_PARK,
     STR_CHEAT_TITLE_RIDE,
 };
+
+static int32_t window_cheats_selectedStaffSpeed = 1;
```
Comment (janisozaur) [inline src/openrct2-ui/windows/Cheats.cpp:358]: This should be moved below the clang format on line

**Comment by KawkMob** [comment]
> I don't like this, there is also the problem that newly created staff will have always the default speed so the user would need to use the cheat again. The right approach is to rework how the cheat works so that newly hired staff will also have the correct speed set, this can be achieved by introducing a new cheat variable rather than modifying the staff entities when the cheat is used. This change just changes the UI aspect of it and I believe people requested this change as they have to constantly use it when hiring new staff.

Yes, you are correct. This change only fixes the problem that was raised here.
To improve the overall behavior that you've described, I see 2 possible approaches:

1. Introduce a getter alongside the existing setter to the low level cheats interface (not sure if possible, but could dive int this) 

2. If not possible, have the values preserved on the OpenRCT2/src/openrct2/Cheats.cpp level so that a getter can be implemented here.

Sorry in advance for an approach that may seem ignorant. This is my first OpenRTC PR, so all input is very welcome!

My suggestion would be to merge the existing fix as it addresses the issue raised by @tonyravioli . 
After that I can look into the the bigger getter mechanism in a separate bug. What do you think @ZehMatt ?

File: src/openrct2-ui/windows/Cheats.cpp (hunk 1 above)
Comment (KawkMob) [reply src/openrct2-ui/windows/Cheats.cpp:359]: Addressed in newest amend.

**Comment by ZehMatt** [comment]
> Yes, you are correct. This change only fixes the problem that was raised here. To improve the overall behavior that you've described, I see 2 possible approaches:
> 
> 1. Introduce a getter alongside the existing setter to the low level cheats interface (not sure if possible, but could dive int this)
> 2. If not possible, have the values preserved on the OpenRCT2/src/openrct2/Cheats.cpp level so that a getter can be implemented here.
> 
> Sorry in advance for an approach that may seem ignorant. This is my first OpenRTC PR, so all input is very welcome!
> 
> My suggestion would be to merge the existing fix as it addresses the issue raised by @tonyravioli . After that I can look into the the bigger getter mechanism in a separate bug. What do you think @ZehMatt ?

1. Would be not possible as the cheat currently has no state, when the cheat is used it just iterates over all staff entities and manipulates some fields that determines the staff speed.
2. That would be the right approach, I would stop modifying the staff entity and instead update the guest/staff code to look at the cheat variable to determine their speed which would also allow the UI to access the current state. Thinking of 3 states the variable can hold, None (Normal Speed), Frozen, Fast, the selection on the UI would basically stay the same just that when Normal is selected it sets SpeedOverride to None.

Given that its just a couple lines I think its okay to be a temporary solution, however the variable name isn't quite right, it should be `_windowCheatsSelectedStaffSpeed`.

**Comment by Broxzier** [comment]
> 2. That would be the right approach, I would stop modifying the staff entity and instead update the guest/staff code to look at the cheat variable to determine their speed which would also allow the UI to access the current state. Thinking of 3 states the variable can hold, None (Normal Speed), Frozen, Fast, the selection on the UI would basically stay the same just that when Normal is selected it sets SpeedOverride to None.

These options (normal, frozen, fast) are essentially energy values. If staff are to check for an enum state, staff that's set to an unusual value (e.g. to freeze them in place) won't stay there when the cheat is used. I'd opt for a `std::optional` that holds the energy behind the cheat, and when clicking `normal` it sets it back to a null-value.

**Comment by ZehMatt** [comment]
> 

That might make it a bit tricky to save, we can just remap enum to Energy level for the cheat.

**Comment by KawkMob** [comment]
> > Yes, you are correct. This change only fixes the problem that was raised here. To improve the overall behavior that you've described, I see 2 possible approaches:
> > 
> > 1. Introduce a getter alongside the existing setter to the low level cheats interface (not sure if possible, but could dive int this)
> > 2. If not possible, have the values preserved on the OpenRCT2/src/openrct2/Cheats.cpp level so that a getter can be implemented here.
> > 
> > Sorry in advance for an approach that may seem ignorant. This is my first OpenRTC PR, so all input is very welcome!
> > My suggestion would be to merge the existing fix as it addresses the issue raised by @tonyravioli . After that I can look into the the bigger getter mechanism in a separate bug. What do you think @ZehMatt ?
> 
> 1. Would be not possible as the cheat currently has no state, when the cheat is used it just iterates over all staff entities and manipulates some fields that determines the staff speed.
> 2. That would be the right approach, I would stop modifying the staff entity and instead update the guest/staff code to look at the cheat variable to determine their speed which would also allow the UI to access the current state. Thinking of 3 states the variable can hold, None (Normal Speed), Frozen, Fast, the selection on the UI would basically stay the same just that when Normal is selected it sets SpeedOverride to None.
> 
> Given that its just a couple lines I think its okay to be a temporary solution, however the variable name isn't quite right, it should be `_windowCheatsSelectedStaffSpeed`.

Name changed. Will raise a new issue including these notes once the change is merged.

**Review by ZehMatt — APPROVED** [review]
Thanks, looks good to me. We might do a release tomorrow (hopefully), I've put this on the 0.4.7 milestone, need to see what the team thinks about merging this for the next release.

**Review by janisozaur — APPROVED** [review]
Approve for 0.4.7

**Comment by KawkMob** [comment]
More comprehensive solution posted at #20756 - covers both scenarios in a less rough way

**Comment by tupaschoal** [comment]
Superseded by #20756

**Pull request closed without merging**